### DISCUSSION TITLE
Changing a dashboard variable refreshes panels that don't reference the variable

### DISCUSSION THREAD

**Discussion by avidhanju** [grafana/grafana#39015] [Dashboards, Panels & Visualizations]
Changing the value of a dashboard variable refreshes panels that don't have a dependency on the variable. It can lead to unnecessary load on the Grafana instance as well as the data source query backend. Plus for complex UI-heavy panels unnecessary redraw happens.

This is very simple to repro. 
In the GIF below I have created a custom variable with two values (Foo, Bar). I also added a TimeSeries panel with Random Walk query type from the -- Grafana -- datasource. There is no reference to the dashboard variable and yet when it changes, the panel refreshes.
![Grafana](https://user-images.githubusercontent.com/10158691/132625849-57932ae9-17c1-4174-aab3-50d74778e1f9.gif)

Expected: Changing the value of a dashboard variable should only update panels that reference the variable.

**Comment by avidhanju** [comment]
@daniellee What are your thoughts on this?

**Comment by mitulk** [comment]
This has multiple drawbacks:
* unnecessary query execution results in load on backend, more chances of throttling
* unnecessary load on client side (rerendering, network traffic, etc.)
* unnecesary cognitive load on user (my panels got re-rendered, let me re grok the charts)

**Comment by daniellee** [comment]
@avidhanju @mitulk Have been talking to some of the engineers working on teams that touch on this feature request. Here's a summary of the discussion.

The initial reaction was that this is hard to do but decided to explore this a bit more. @hugohaggmark has previously worked on the variable system. Thanks to that work, we do know exactly where all the variables are used due to the variable inspection feature.

We could probably do a smarter update when a variable changes, at least in theory. But the challenge is how to publish the affected panels only once when you have chained variables (cascading).

Hugo is working on a proof of concept and has a rough draft here: https://github.com/grafana/grafana/pull/39420

We will get back with more details once we evaluate this approach and how it fits into our current backlog.

**Reply by daniellee** [reply]
@hugohaggmark had a question come about this today - will this result in some panels being slightly out of date? For example, if I have two panels A and B, and A uses a variable and B does not. When I change the variable, then A will get updated but not B and B's data might then be stale.

Any thoughts on whether that is a problem? Would we need to disable this logic if the time range is relative (Last 15 minutes or similar)?

**Reply by hugohaggmark** [reply]
Hmm, that makes it more complicated. I'm a bit worried we add this update on change complexity and in the end it won't really be used because of various reasons like this one.

**Reply by daniellee** [reply]
Talked to Hugo and we decided to start with make this an opt-in feature. Hugo will add a dashboard setting for enabling this logic. If a user is using a data source like Prometheus where queries return quickly and where users often look at data as it is being ingested, then it it seems unlikely that they would need this feature.

**Reply by avidhanju** [reply]
@daniellee I think it depends on how you expect the users to think about stale panels. In your example panel B was stale because relative time changed and not because a variable that panel B uses changed. In other words, the variable change is not the reason for _stale_ data in panel B but rather use of relative time range and other panels that updated due to variable change using the newer time range causes that perception. A semi-reasonable expectation can be to ask the customers to use the refresh button to update the _stale_ panels but I can understand how an extra click can be jarring for many users.

Instead of putting this new feature behind a feature flag, is it possible to expose a new property on variables called something like "strict mode"? If dashboard authors choose to turn on the strict mode for a variable, that variable will only update panels that directly use it (and we can ignore this mode for nested variables) and all other variables behave the way they currently do. This way, it is an opt in feature on a per variable basis. @hugohaggmark Thoughts?

**Reply by hugohaggmark** [reply]
Hi @avidhanju, as a first MVP I would like to add a dashboard property that is shown when you edit all the variables. It will be a toggle/switch/checkbox where you can opt in for `strict mode` for the whole dashboard. Doing this per variable or per panel could be a future improvement, but we'd like to collect some community feedback before introducing this complexity. Does that sound Ok?

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
